File: go/Zadanie6/hash_Table.go
// hash_Table.go
package main

import (
	"math"
	"strconv"
)

type CollisionResolution int

const (
	CHAINING CollisionResolution = iota
	LINEAR_PROBING
	QUADRATIC_PROBING
)

type HashTable struct {
	table  *Array
	size   int
	method CollisionResolution
	prime  int
}

func NewHashTable(tableSize int, resolutionMethod CollisionResolution) *HashTable {
	ht := &HashTable{
		table:  NewArray(tableSize),
		size:   tableSize,
		method: resolutionMethod,
	}

	ht.prime = ht.getNextPrime(tableSize / 2)

	if resolutionMethod == CHAINING {
		for i := 0; i < tableSize; i++ {
			// Убрана неиспользуемая переменная newList
			ht.table.Add("")
		}
	} else {
		for i := 0; i < tableSize; i++ {
			ht.table.Add("")
		}
	}

	return ht
}

func (ht *HashTable) hash1(key int) int {
	return int(math.Abs(float64(key))) % ht.size
}

func (ht *HashTable) hash2(key int) int {
	return ht.prime - (int(math.Abs(float64(key))) % ht.prime)
}

func (ht *HashTable) getNextPrime(n int) int {
	if n < 2 {
		return 2
	}

	for {
		isPrime := true
		for i := 2; i <= int(math.Sqrt(float64(n))); i++ {
			if n%i == 0 {
				isPrime = false
				break
			}
		}
		if isPrime {
			return n
		}
		n++
	}
}

func (ht *HashTable) findSlotOpenAddressing(key int) int {
	index := ht.hash1(key)
	keyStr := strconv.Itoa(key)

	if ht.method == LINEAR_PROBING {
		for i := 0; i < ht.size; i++ {
			currentIndex := (index + i) % ht.size
			currentValue := ht.table.Get(currentIndex)

			if ht.isEmpty(currentIndex) || ht.isDeleted(currentIndex) {
				return currentIndex
			}

			if currentValue == keyStr {
				return currentIndex
			}
		}
	} else if ht.method == QUADRATIC_PROBING {
		for i := 0; i < ht.size; i++ {
			currentIndex := (index + i*i) % ht.size
			currentValue := ht.table.Get(currentIndex)

			if ht.isEmpty(currentIndex) || ht.isDeleted(currentIndex) {
				return currentIndex
			}

			if currentValue == keyStr {
				return currentIndex
			}
		}
	}

	return -1
}

func (ht *HashTable) isDeleted(index int) bool {
	if ht.method == CHAINING {
		return false
	}
	value := ht.table.Get(index)
	return value == "DELETED"
}

func (ht *HashTable) isEmpty(index int) bool {
	if ht.method == CHAINING {
		// Для цепочек всегда возвращаем false, так как логика другая
		return false
	} else {
		value := ht.table.Get(index)
		return value == "" || value == "DELETED"
	}
}

func (ht *HashTable) Insert(key int) {
	if ht.method == CHAINING {
		// В этой упрощенной версии для цепочек используем массив
		index := ht.hash1(key)
		keyStr := strconv.Itoa(key)
		current := ht.table.Get(index)

		if current == "" {
			ht.table.Set(index, keyStr)
		} else {
			// Для простоты добавляем через разделитель
			ht.table.Set(index, current+","+keyStr)
		}
	} else {
		index := ht.findSlotOpenAddressing(key)
		if index != -1 {
			ht.table.Set(index, strconv.Itoa(key))
		}
	}
}

func (ht *HashTable) Search(key int) bool {
	keyStr := strconv.Itoa(key)

	if ht.method == CHAINING {
		index := ht.hash1(key)
		value := ht.table.Get(index)

		if value == "" {
			return false
		}

		// Простая проверка для цепочек
		if value == keyStr {
			return true
		}

		// Проверяем в цепочке (разделенной запятыми)
		// В реальной реализации нужно использовать связанный список
		return containsValue(value, keyStr)
	} else {
		index := ht.findSlotOpenAddressing(key)
		return index != -1 && ht.table.Get(index) == keyStr
	}
}

func (ht *HashTable) Remove(key int) {
	keyStr := strconv.Itoa(key)

	if ht.method == CHAINING {
		index := ht.hash1(key)
		value := ht.table.Get(index)

		if value != "" {
			if value == keyStr {
				ht.table.Set(index, "")
			} else {
				// Упрощенное удаление из цепочки
				newValue := removeValue(value, keyStr)
				ht.table.Set(index, newValue)
			}
		}
	} else {
		index := ht.findSlotOpenAddressing(key)
		if index != -1 && ht.table.Get(index) == keyStr {
			ht.table.Set(index, "DELETED")
		}
	}
}

func (ht *HashTable) GetStatistics() (int, int, float64) {
	minLength := int(^uint(0) >> 1) // Max int
	maxLength := 0
	totalElements := 0
	nonEmptyBuckets := 0

	if ht.method == CHAINING {
		for i := 0; i < ht.size; i++ {
			value := ht.table.Get(i)
			if value != "" {
				chainLength := countChainLength(value)
				if chainLength < minLength {
					minLength = chainLength
				}
				if chainLength > maxLength {
					maxLength = chainLength
				}
				totalElements += chainLength
				nonEmptyBuckets++
			}
		}
	} else {
		visited := make([]bool, ht.size)

		for i := 0; i < ht.size; i++ {
			if !ht.isEmpty(i) && !ht.isDeleted(i) && !visited[i] {
				clusterSize := 0
				current := i

				for !ht.isEmpty(current) && !visited[current] {
					if !ht.isDeleted(current) {
						clusterSize++
					}
					visited[current] = true
					current = (current + 1) % ht.size
				}

				if clusterSize > 0 {
					if clusterSize < minLength {
						minLength = clusterSize
					}
					if clusterSize > maxLength {
						maxLength = clusterSize
					}
					totalElements += clusterSize
					nonEmptyBuckets++
				}
			}
		}
	}

	if minLength == int(^uint(0)>>1) {
		minLength = 0
	}

	avgLength := 0.0
	if nonEmptyBuckets > 0 {
		avgLength = float64(totalElements) / float64(nonEmptyBuckets)
	}

	return minLength, maxLength, avgLength
}

func (ht *HashTable) PrintDetailedStatistics() {
	totalElements := 0
	emptyBuckets := 0
	maxChain := 0
	minChain := int(^uint(0) >> 1)

	print("Метод разрешения коллизий: ")
	switch ht.method {
	case CHAINING:
		println("Метод цепочек")
	case LINEAR_PROBING:
		println("Линейное пробирование")
	case QUADRATIC_PROBING:
		println("Квадратичное пробирование")
	}

	println("Распределение:")

	for i := 0; i < ht.size; i++ {
		chainLength := 0

		if ht.method == CHAINING {
			value := ht.table.Get(i)
			if value != "" {
				chainLength = countChainLength(value)
			}
		} else {
			if !ht.isEmpty(i) && !ht.isDeleted(i) {
				chainLength = 1
			}
		}

		if chainLength == 0 {
			emptyBuckets++
		}
		if chainLength > maxChain {
			maxChain = chainLength
		}
		if chainLength < minChain && chainLength > 0 {
			minChain = chainLength
		}

		totalElements += chainLength

		if i < 10 {
			printf("Ячейка %d: %d элементов\n", i, chainLength)
		}
	}

	println("\nОбщая статистика:")
	if minChain == int(^uint(0)>>1) {
		minChain = 0
	}
	printf("Минимальная длина непустой цепочки: %d\n", minChain)
	printf("Максимальная длина цепочки: %d\n", maxChain)
	printf("Пустых ячеек: %d из %d\n", emptyBuckets, ht.size)
	printf("Общее количество элементов: %d\n", totalElements)
	printf("Коэффициент заполнения: %.2f\n", float64(totalElements)/float64(ht.size))
}

func (ht *HashTable) SetResolutionMethod(method CollisionResolution) {
	ht.method = method
}

// Вспомогательные функции
func containsValue(chain string, value string) bool {
	// Простая проверка для демонстрации
	return chain == value
}

func removeValue(chain string, value string) string {
	// Упрощенная реализация для демонстрации
	if chain == value {
		return ""
	}
	return chain
}

func countChainLength(chain string) int {
	if chain == "" {
		return 0
	}
	// Упрощенный подсчет - в реальной реализации нужно разбирать цепочку
	return 1
}

func printf(format string, args ...interface{}) {
	// Простая реализация форматированного вывода
	print(format)
}

File: с++/Zadanie6/Zadanie6/hash_table.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <cmath>
#include "hash_table.h"

using namespace std;

// Основная хеш-функция - метод деления
int HashTable::hash1(int key) const {
    return abs(key) % size;
}

// Вторая хеш-функция для двойного хеширования
int HashTable::hash2(int key) const {
    return prime - (abs(key) % prime);
}

// Получение следующего простого числа после n
int HashTable::getNextPrime(int n) const {
    while (true) {
        bool isPrime = true;
        // Проверка на простоту
        for (int i = 2; i <= sqrt(n); i++) {
            if (n % i == 0) {
                isPrime = false;
                break;
            }
        }
        if (isPrime) return n;
        n++;
    }
}

// Методы для работы с цепочками (chaining)

// Получить список по индексу таблицы
ForwardListTwo* HashTable::getList(int index) const {
    if (index < 0 || index >= size) return nullptr;
    string ptrStr = get(*table, index);
    if (ptrStr.empty()) return nullptr;
    // Преобразование строки обратно в указатель
    return reinterpret_cast<ForwardListTwo*>(stoll(ptrStr));
}

// Очистка списка и освобождение памяти
void HashTable::clearList(ForwardListTwo* list) {
    if (!list) return;
    // Удаление всех узлов списка
    while (list->head != nullptr) {
        deleteNodeHeadTwo(list);
    }
    delete list;
}

// Методы для открытой адресации

// Проверка, помечена ли ячейка как удаленная
bool HashTable::isDeleted(int index) const {
    if (method == CollisionResolution::CHAINING) return false;
    string value = get(*table, index);
    return value == "DELETED";
}

// Проверка, пуста ли ячейка
bool HashTable::isEmpty(int index) const {
    if (method == CollisionResolution::CHAINING) {
        // Для метода цепочек - проверяем пустоту списка
        ForwardListTwo* list = getList(index);
        return !list || countNodesTwo(*list) == 0;
    }
    else {
        // Для открытой адресации - проверяем значение ячейки
        string value = get(*table, index);
        return value.empty() || value == "DELETED";
    }
}

// Поиск слота для вставки/поиска при открытой адресации
int HashTable::findSlotOpenAddressing(int key) const {
    int index = hash1(key);
    string keyStr = to_string(key);

    // Линейное пробирование
    if (method == CollisionResolution::LINEAR_PROBING) {
        for (int i = 0; i < size; i++) {
            int currentIndex = (index + i) % size;
            string currentValue = get(*table, currentIndex);

            // Пустая ячейка или удаленная - можем использовать
            if (isEmpty(currentIndex) || isDeleted(currentIndex)) {
                return currentIndex;
            }
            // Нашли ключ - возвращаем для поиска/удаления
            if (currentValue == keyStr) {
                return currentIndex;
            }
        }
    }
    // Квадратичное пробирование
    else if (method == CollisionResolution::QUADRATIC_PROBING) {
        for (int i = 0; i < size; i++) {
            int currentIndex = (index + i * i) % size;
            string currentValue = get(*table, currentIndex);

            if (isEmpty(currentIndex) || isDeleted(currentIndex)) {
                return currentIndex;
            }
            if (currentValue == keyStr) {
                return currentIndex;
            }
        }
    }
    return -1; // Таблица заполнена
}

// Конструктор хеш-таблицы
HashTable::HashTable(int tableSize, CollisionResolution resolutionMethod)
    : size(tableSize), method(resolutionMethod) {
    table = new Array;
    init(*table, size);
    prime = getNextPrime(size / 2); // Для двойного хеширования

    // Инициализация в зависимости от метода разрешения коллизий
    if (method == CollisionResolution::CHAINING) {
        // Инициализация цепочек - создаем список для каждой ячейки
        for (int i = 0; i < size; i++) {
            ForwardListTwo* newList = new ForwardListTwo;
            newList->head = nullptr;
            newList->tail = nullptr;
            add(*table, "");
            // Сохраняем указатель на список как строку
            set(*table, i, to_string(reinterpret_cast<long long>(newList)));
        }
    }
    else {
        // Инициализация для открытой адресации - пустые ячейки
        for (int i = 0; i < size; i++) {
            add(*table, ""); // Пустые ячейки
        }
    }
}

// Деструктор хеш-таблицы
HashTable::~HashTable() {
    // Очистка памяти в зависимости от метода
    if (method == CollisionResolution::CHAINING) {
        for (int i = 0; i < size; i++) {
            ForwardListTwo* list = getList(i);
            clearList(list);
        }
    }
    destroy(*table);
    delete table;
}

// Вставка элемента в хеш-таблицу
void HashTable::insert(int key) {
    if (method == CollisionResolution::CHAINING) {
        // Метод цепочек - добавляем в конец списка
        int index = hash1(key);
        ForwardListTwo* list = getList(index);
        if (list) {
            addNodeTailTwo(list, to_string(key));
        }
    }
    else {
        // Открытая адресация - находим свободный слот
        int index = findSlotOpenAddressing(key);
        if (index != -1) {
            set(*table, index, to_string(key));
        }
    }
}

// Поиск элемента в хеш-таблице
bool HashTable::search(int key) const {
    if (method == CollisionResolution::CHAINING) {
        // Метод цепочек - ищем в списке
        int index = hash1(key);
        ForwardListTwo* list = getList(index);
        if (!list) return false;
        return findNodeIndexTwo(list, to_string(key));
    }
    else {
        // Открытая адресация - используем поиск слота
        int index = findSlotOpenAddressing(key);
        return index != -1 && get(*table, index) == to_string(key);
    }
}

// Удаление элемента из хеш-таблицы
void HashTable::remove(int key) {
    if (method == CollisionResolution::CHAINING) {
        // Метод цепочек - удаляем из списка
        int index = hash1(key);
        ForwardListTwo* list = getList(index);
        if (list) {
            deleteNodeIndexTwo(list, to_string(key));
        }
    }
    else {
        // Открытая адресация - помечаем как удаленное
        int index = findSlotOpenAddressing(key);
        if (index != -1 && get(*table, index) == to_string(key)) {
            set(*table, index, "DELETED");
        }
    }
}

// Получение статистики хеш-таблицы
void HashTable::getStatistics(int& minLength, int& maxLength, double& avgLength) const {
    minLength = INT_MAX;
    maxLength = 0;
    int totalElements = 0;
    int nonEmptyBuckets = 0;

    if (method == CollisionResolution::CHAINING) {
        // Логика для метода цепочек
        for (int i = 0; i < size; i++) {
            ForwardListTwo* list = getList(i);
            if (list) {
                int listSize = countNodesTwo(*list);
                if (listSize > 0) {
                    if (listSize < minLength) minLength = listSize;
                    if (listSize > maxLength) maxLength = listSize;
                    totalElements += listSize;
                    nonEmptyBuckets++;
                }
            }
        }
    }
    else {
        // Логика для открытой адресации - анализ кластеров
        Array visitedArray;
        init(visitedArray, size);

        // Инициализируем массив посещенных ячеек
        for (int i = 0; i < size; i++) {
            add(visitedArray, "0"); // "0" = false, "1" = true
        }

        // Анализ кластеров для линейного пробирования
        for (int i = 0; i < size; i++) {
            if (!isEmpty(i) && !isDeleted(i) && get(visitedArray, i) == "0") {
                int clusterSize = 0;
                int current = i;

                // Измеряем размер кластера
                while (!isEmpty(current) && get(visitedArray, current) == "0") {
                    if (!isDeleted(current)) {
                        clusterSize++;
                    }
                    set(visitedArray, current, "1"); // помечаем как посещенный
                    current = (current + 1) % size; // для линейного пробирования
                }

                if (clusterSize > 0) {
                    if (clusterSize < minLength) minLength = clusterSize;
                    if (clusterSize > maxLength) maxLength = clusterSize;
                    totalElements += clusterSize;
                    nonEmptyBuckets++;
                }
            }
        }

        destroy(visitedArray);
    }

    // Корректировка значений по умолчанию
    if (minLength == INT_MAX) minLength = 0;
    avgLength = nonEmptyBuckets > 0 ? static_cast<double>(totalElements) / nonEmptyBuckets : 0;
}

// Вывод подробной статистики хеш-таблицы
void HashTable::printDetailedStatistics() const {
    int totalElements = 0;
    int emptyBuckets = 0;
    int maxChain = 0;
    int minChain = INT_MAX;

    // Вывод информации о методе разрешения коллизий
    cout << "Метод разрешения коллизий: ";
    switch (method) {
    case CollisionResolution::CHAINING:
        cout << "Метод цепочек\n";
        break;
    case CollisionResolution::LINEAR_PROBING:
        cout << "Линейное пробирование\n";
        break;
    case CollisionResolution::QUADRATIC_PROBING:
        cout << "Квадратичное пробирование\n";
        break;
    }

    cout << "Распределение:\n";

    // Сбор статистики по всем ячейкам
    for (int i = 0; i < size; i++) {
        int chainLength = 0;

        if (method == CollisionResolution::CHAINING) {
            ForwardListTwo* list = getList(i);
            chainLength = list ? countNodesTwo(*list) : 0;
        }
        else {
            chainLength = (!isEmpty(i) && !isDeleted(i)) ? 1 : 0;
        }

        // Обновление статистики
        if (chainLength == 0) emptyBuckets++;
        if (chainLength > maxChain) maxChain = chainLength;
        if (chainLength < minChain && chainLength > 0) minChain = chainLength;

        totalElements += chainLength;

        // Вывод первых 10 ячеек для примера
        if (i < 10) {
            cout << "Ячейка " << i << ": " << chainLength << " элементов\n";
        }
    }

    // Вывод итоговой статистики
    cout << "\nОбщая статистика:\n";
    cout << "Минимальная длина непустой цепочки: " << (minChain == INT_MAX ? 0 : minChain) << "\n";
    cout << "Максимальная длина цепочки: " << maxChain << "\n";
    cout << "Пустых ячеек: " << emptyBuckets << " из " << size << "\n";
    cout << "Общее количество элементов: " << totalElements << "\n";
    cout << "Коэффициент заполнения: " << static_cast<double>(totalElements) / size << "\n";
}

// Установка метода разрешения коллизий
void HashTable::setResolutionMethod(CollisionResolution method) {
    this->method = method;
}